utils_cv/detection/model.py

```diff
@@ -0,0 +1,283 @@
+from torchvision.models.detection.faster_rcnn import FastRCNNPredictor
+from torch.utils.data import Dataset, DataLoader
+import matplotlib.pyplot as plt
+
+from .references.transforms import RandomHorizontalFlip, Compose, ToTensor
+from .references.engine import train_one_epoch, evaluate
+from .references.utils import collate_fn
+from .references.coco_eval import CocoEvaluator
+from .plot import BoundingBoxParams, plot_boxes, plot_grid
+
+
+def get_bounding_boxes(
+    pred: List[dict], threshold: int = 0.6
+) -> Tuple[List[int], List[List[int]]]:
+    """ Gets the bounding boxes and labels from a prediction given a threshold.
+
+    Args:
+        pred: the output of passing in an image to torchvision's FasterRCNN
+        model
+        threshold: the minimum threshold to accept.
+
+    Return:
+        a list of labels and bounding boxes that pass the minimum threshold.
+    """
+    pred_labels = list(pred[0]["labels"].cpu().numpy())
+    pred_boxes = list(pred[0]["boxes"].detach().cpu().numpy().astype(np.int32))
+    pred_scores = list(pred[0]["scores"].cpu().numpy())
+
+    qualified_labels = []
+    qualified_boxes = []
+    for label, box, score in zip(pred_labels, pred_boxes, pred_scores):
+        if score > threshold:
+            qualified_labels.append(label)
+            qualified_boxes.append(box)
+    return qualified_labels, qualified_boxes
+
+
+def get_transform(train: bool) -> List[object]:
+    """ Gets basic the transformations to apply to images.
+
+    Args:
+        train: whether or not we are getting transformations for the training
+        set.
+
+    Returns:
+        A list of transforms to apply.
+    """
+    transforms = []
+    transforms.append(ToTensor())
+    if train:
+        transforms.append(RandomHorizontalFlip(0.5))
+    return Compose(transforms)
+
+
+def get_pretrained_fasterrcnn(num_classes: int) -> nn.Module:
+    """ Gets a pretrained FasterRCNN model
+
+    Args:
+        num_classes: the number of classes to be detected
+
+    Returns
+        The model to fine-tine/inference with
+    """
+    # load a model pre-trained pre-trained on COCO
+    model = fasterrcnn_resnet50_fpn(pretrained=True)
+
+    # get number of input features for the classifier
+    in_features = model.roi_heads.box_predictor.cls_score.in_features
+
+    # replace the pre-trained head with a new one
+    # that has num_classes which is based on the dataset
+    model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes)
+    return model
+
+
+class DetectionLearner:
+    """ Detection Learner for Object Detection"""
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        lr: float,
+        train_ratio: float = 0.8,
+        batch_size: int = 2,
+        model: nn.Module = None,
+        momentum: float = 0.9,
+        weight_decay: float = 0.0005,
+    ):
+        """ Initialize leaner object. """
+        self.device = (
+            torch.device("cuda")
```

Comment:
Could make a helper function out of this called e.g. "def torch_device()" if this phrase is used more often.